### PR TITLE
[BEAM-5816] Finish Flink bundles exactly once

### PR DESCRIPTION
**Pull request by mxm**
When an error an occurred in an operator chain, `dipose()` could be called
multiple times leading to duplicate calls to `invokeFinishBundle()`, some of
which actions expect to release resources only once. Also, the bundle expiration
timer could call finish bundle simultaneously with `dipose()`.

This makes the finish bundle logic execute exactly once by making it
atomic. Note: This did not affect correctness of results but the change ensures
there are no duplicate cleanup calls.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by mxm** [comment]
CC @mwylde @tweise

**Comment by tweise** [comment]
Why do we need to finish bundles in dispose? Shouldn't bundles be finished as part of regular processing only. What happens when the operator is removed after an exception? Are we going to "finish" partially processed bundles?

**Comment by mxm** [comment]
There is nothing about the general bundle logic that changed in this PR. The commit which introduced the finish bundle call before the teardown does not explain why it was necessary. Nevertheless, it seems reasonable to try to finish a pending bundle before tearing down the DoFn. The bundle is finished in dispose() because there it is guaranteed to be called because in case of errors close() will not be called.

>What happens when the operator is removed after an exception? Are we going to "finish" partially processed bundles?

It is ok to finish bundles even when there are exceptions. Bundles can be of arbitrary element size.

**Comment by mxm** [comment]
Run Flink ValidatesRunner

**Comment by mxm** [comment]
Is there anything blocking this PR?

**Comment by tweise** [comment]
So I actually don't see any reason to finish bundles after execution has already stopped (for example as result of pipeline cancellation). I can see that bundles need to be finished after final watermark and on checkpoint boundary, but not during teardown. Can we remove this from dispose entirely?

**Comment by mxm** [comment]
Fair point. I've moved `finishBundle` to the `close()` method.

**Comment by mxm** [comment]
Run Flink ValidatesRunner

**Comment by mxm** [comment]
Run Flink ValidatesRunner

**Comment by mxm** [comment]
Run Java Flink PortableValidatesRunner Streaming

**Comment by mxm** [comment]
@tweise Should be good to go

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java [hunk 1]
```diff
@@ -178,7 +179,7 @@
   private transient PushedBackElementsHandler<WindowedValue<InputT>> pushedBackElementsHandler;
 
   // bundle control
-  private transient boolean bundleStarted = false;
+  private transient AtomicBoolean bundleStarted;
```
Comment (tweise) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java]: Is there a scenario of multiple threads accessing this?

**Comment by tweise** [comment]
I think there are two things we want to accomplish:
1) After an exception or cancel, we don't want any more "normal" processing (looks like this is now addressed)
2) Best effort resource cleanup for user code (not sure about that one, would this happen after an exception)?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java (hunk 1 above)
Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java]: Yes, there is a timer thread for the "finish bundle by time" feature, see `FlinkPipelineOptions#setMaxBundleTimeMills`.

**Comment by mxm** [comment]
>    1. After an exception or cancel, we don't want any more "normal" processing (looks like this is now addressed)

Yes, this is addressed. Bundle will only be finalized in case of normal operation. Though my original intention here was to fix the race condition between the timer thread for finishing the bundle and a regular bundle finish

>    2. Best effort resource cleanup for user code (not sure about that one, would this happen after an exception)?

Cleanup is performed via DoFn#teardown which will be called in any case in `dispose()`.

**Pull request merged**